pkg/urcrypt/urcrypt/pedersen.c

```diff
@@ -0,0 +1,264 @@
+    0x4e, 0x62, 0x10, 0x62, 0xc0, 0xe0, 0x56, 0xc1,
+    0xd0, 0x40, 0x5d, 0x26, 0x6e, 0x10, 0x26, 0x8a
+  };
+
+  const u8 p1_buf[] = {
+    // x coordinate
+    0x02, 0x34, 0x28, 0x7d, 0xcb, 0xaf, 0xfe, 0x7f,
+    0x96, 0x9c, 0x74, 0x86, 0x55, 0xfc, 0xa9, 0xe5,
+    0x8f, 0xa8, 0x12, 0x0b, 0x6d, 0x56, 0xeb, 0x0c,
+    0x10, 0x80, 0xd1, 0x79, 0x57, 0xeb, 0xe4, 0x7b,
+
+    // y coordinate
+    0x03, 0xb0, 0x56, 0xf1, 0x00, 0xf9, 0x6f, 0xb2,
+    0x1e, 0x88, 0x95, 0x27, 0xd4, 0x1f, 0x4e, 0x39,
+    0x94, 0x01, 0x35, 0xdd, 0x7a, 0x6c, 0x94, 0xcc,
+    0x6e, 0xd0, 0x26, 0x8e, 0xe8, 0x9e, 0x56, 0x15
+  };
+
+  const u8 p2_buf[] = {
+    // x coordinate
+    0x04, 0xfa, 0x56, 0xf3, 0x76, 0xc8, 0x3d, 0xb3,
+    0x3f, 0x9d, 0xab, 0x26, 0x56, 0x55, 0x8f, 0x33,
+    0x99, 0x09, 0x9e, 0xc1, 0xde, 0x5e, 0x30, 0x18,
+    0xb7, 0xa6, 0x93, 0x2d, 0xba, 0x8a, 0xa3, 0x78,
+
+    // y coordinate
+    0x03, 0xfa, 0x09, 0x84, 0xc9, 0x31, 0xc9, 0xe3,
+    0x81, 0x13, 0xe0, 0xc0, 0xe4, 0x7e, 0x44, 0x01,
+    0x56, 0x27, 0x61, 0xf9, 0x2a, 0x7a, 0x23, 0xb4,
+    0x51, 0x68, 0xf4, 0xe8, 0x0f, 0xf5, 0xb5, 0x4d
+  };
+
+  const u8 p3_buf[] = {
+    // x coordinate
+    0x04, 0xba, 0x4c, 0xc1, 0x66, 0xbe, 0x8d, 0xec,
+    0x76, 0x49, 0x10, 0xf7, 0x5b, 0x45, 0xf7, 0x4b,
+    0x40, 0xc6, 0x90, 0xc7, 0x47, 0x09, 0xe9, 0x0f,
+    0x3a, 0xa3, 0x72, 0xf0, 0xbd, 0x2d, 0x69, 0x97,
+
+    // y coordinate
+    0x00, 0x40, 0x30, 0x1c, 0xf5, 0xc1, 0x75, 0x1f,
+    0x4b, 0x97, 0x1e, 0x46, 0xc4, 0xed, 0xe8, 0x5f,
+    0xca, 0xc5, 0xc5, 0x9a, 0x5c, 0xe5, 0xae, 0x7c,
+    0x48, 0x15, 0x1f, 0x27, 0xb2, 0x4b, 0x21, 0x9c
+  };
+
+  const u8 p4_buf[] = {
+    // x coordinate
+    0x05, 0x43, 0x02, 0xdc, 0xb0, 0xe6, 0xcc, 0x1c,
+    0x6e, 0x44, 0xcc, 0xa8, 0xf6, 0x1a, 0x63, 0xbb,
+    0x2c, 0xa6, 0x50, 0x48, 0xd5, 0x3f, 0xb3, 0x25,
+    0xd3, 0x6f, 0xf1, 0x2c, 0x49, 0xa5, 0x82, 0x02,
+
+    // y coordinate
+    0x01, 0xb7, 0x7b, 0x3e, 0x37, 0xd1, 0x35, 0x04,
+    0xb3, 0x48, 0x04, 0x62, 0x68, 0xd8, 0xae, 0x25,
+    0xce, 0x98, 0xad, 0x78, 0x3c, 0x25, 0x56, 0x1a,
+    0x87, 0x9d, 0xcc, 0x77, 0xe9, 0x9c, 0x24, 0x26
+  };
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p0), p0_buf, 64, crv);
+  if (ret != 0) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p1), p1_buf, 64, crv);
+  if (ret != 0) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p2), p2_buf, 64, crv);
+  if (ret != 0) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p3), p3_buf, 64, crv);
+  if (ret != 0) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p4), p4_buf, 64, crv);
+  return ret;
+}
+
+int
+urcrypt_pedersen(uint8_t *a, size_t a_len, uint8_t *b, size_t b_len, uint8_t out[32])
+{
+  int ret = 0;
+  urcrypt__reverse(a_len, a);
+  urcrypt__reverse(b_len, b);
+  if (a_len == 32) {
+    a[0] = a[0] & 0x0f;
+  }
+  if (b_len == 32) {
+    b[0] = b[0] & 0x0f;
+  }
+
+  if (curve_params == NULL) {
+    curve_params = (ec_params *)malloc(sizeof(ec_params));
```

Comment:
make some kind of urcrypt_pederson_init function (see urcrypt_secp_init and where it is called in vere), and don't use malloc at all.